src/design_notes/general_coroutines.md

```diff
@@ -0,0 +1,258 @@
+  values manually wrapped in `GeneratorState`.
+- Enum-wrapping here is analogous to Ok-wrapping elsewhere. Similar debates
+  result.
+- When using enum-wrapping, the syntax to specify distinct return/yield types is
+  hotly debated.
+- The difference is strictly one of ergonomics.
+  - Keep in mind any un-ergonomic cases can potentially be tackled with more
+    specialized, macro-powered syntaxes.
+  - Auto-enum-wrapping can slightly improve type safety in some cases where
+    `return` should be treated specially to avoid bugs.
+  - No-enum-wrapping when combined with the `impl Fn*` choice of trait, allow
+    the coroutine syntax to be used directly with existing higher-order methods
+    on iterator, stream, collection types, async traits, etc.
+
+
+## Movability
+
+- All proposals want movability/`impl Unpin` to be inferred.
+  - Exact inference rules may only be revealed by an attempt at implementation.
+- Soundness of `pin_mut!` is a little tricky but seems to be fine no matter what.
+  - If the resulting mutable reference is live across a yield ⇒ coroutine is
+    `!Unpin` because of inference rules
+  - If the pinned data is `!Unpin` and dropped after a yield ⇒ coroutine is
+    `!Unpin` because witness contains `!Unpin` data
+  - Thus, if the coroutine can be moved after resume, any data stack-pinned
+    (really witness-pinned) by `pin_mut!` is not referenced and is `Unpin`.
+- Until inference is solved, the `static` keyword can be used as a modifier.
+
+## "Once" coroutines
+
+- A lot of coroutines consume captured data.
+- These coroutines (notably futures) can be resumed many times but can only be
+  run through "once".
+- In contrast to non-yield `FnOnce` closures, this can not be solved at the type
+  level because a coroutine can run out after an arbitrary, runtime-dependent
+  number of resumptions.
+  - Attempts to discriminate with enums tend to run up against `Pin`.
+- Coroutines must have the ability to block restart with a `panic!`.
+  - Following `return`.
+  - Following `panic!` and recovery.
+  - The term "poison state" technically refers to only the later case. But here
+    I will use it to mean any state from which the closure can not be resumed.
+- [RFC-2781][2] and [eRFC-2033][1] propose that all coroutines become poisoned
+  after returning. Always.
+- [MCP-49][3] optionally proposes that closures can only be poisoned if
+  explicitly annotated, and otherwise loop around after return, yield-or-no.
+  - The looping semantics can be very handy in a few situations. The work around
+    is to `return => loop { yield; continue; }`
+  - But the behavior of the `mut` modifier may be too obscure and require too
+    much explanation vs "closures poison if they contain yield".
+
+## Async coroutines
+
+- *Async generators must totally be a thing.* But any specialized generator
+  syntax is a distinct discussion which needs to solve totally different
+  problems.
```

Comment:
Personally, I'd just as soon get yield closures out the door first and then leave the generator syntax to the crates. But I actually think it would be very reasonable to include both a dedicated generator syntax and yield closures in the long term because they are each fairly simple, narrow features with totally independent motivation:
- Users want to implement `Iterator` & `Stream` very concisely => generators
- Users want a general-purpose coroutine primitive => yield closures
- Users want an easier way to express complex statefulness in `FnMut` closures => yield closures

What I don't want is a whole new coroutine syntax that tries to be the generator syntax but with resume args. That would be difficult, ugly, and silly. Far better to find a related feature (closures) and minimally extend it to make arbitrary use of the coroutine transform (yield).

I'll try to elaborate more on that in the notes. Or at least try to make that point more clearly.

Reply:
> `fn next_fn<T>`

Just to note, this exists already in the stdlib as [`std::iter::from_fn`](https://doc.rust-lang.org/std/iter/fn.from_fn.html). So for an `Unpin` `yield` closure, you could just use it to write iterators.